### PR TITLE
Better OpenAPI spec v3 support: allOf, anyOf, oneOf

### PR DESCRIPTION
**Pull request by wing328**
### PR checklist

- [ ] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ ] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [ ] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `3.4.x`, `4.0.x`. Default: `master`.
- [ ] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

- fix composition
- fix inheritance support
- add mustache tags `allOf`, `anyOf`, `oneOf`
- add tests for discriminator, allOf, anyOf, etc in Rub client generator
- update Ruby templates to support inheritance, anyOf, oneOf, etc

Related issues/PRs
- https://github.com/OpenAPITools/openapi-generator/issues/1358
- https://github.com/OpenAPITools/openapi-generator/issues/1206
- https://github.com/OpenAPITools/openapi-generator/issues/927
- https://github.com/OpenAPITools/openapi-generator/issues/1068
- https://github.com/OpenAPITools/openapi-generator/pull/1099
- https://github.com/OpenAPITools/openapi-generator/issues/634
- https://github.com/OpenAPITools/openapi-generator/issues/1125
- https://github.com/OpenAPITools/openapi-generator/issues/537

cc @OpenAPITools/generator-core-team

### CODE REVIEW

**Comment by etherealjoy** [comment]
also this one #537?

**Comment by wing328** [comment]
@etherealjoy 👍 updated the list

File: samples/schema/petstore/mysql/mysql_schema.sql [hunk 1]
```diff
@@ -95,6 +95,7 @@ CREATE TABLE IF NOT EXISTS `Capitalization` (
 --
 
 CREATE TABLE IF NOT EXISTS `Cat` (
+  `declawed` TINYINT(1) DEFAULT NULL,
```
Comment (ybelenko) [inline samples/schema/petstore/mysql/mysql_schema.sql]: Looks like `{{{#vars}}}{{{/vars}}}` array now has duplicated values. It breaks current implementation.

File: samples/schema/petstore/mysql/mysql_schema.sql [hunk 2]
```diff
@@ -131,6 +132,7 @@ CREATE TABLE IF NOT EXISTS `Client` (
 --
 
 CREATE TABLE IF NOT EXISTS `Dog` (
+  `breed` TEXT DEFAULT NULL,
```
Comment (ybelenko) [inline samples/schema/petstore/mysql/mysql_schema.sql]: Same here.

File: samples/server/petstore/php-slim/lib/Model/Cat.php [hunk 3]
```diff
@@ -10,6 +10,9 @@
 class Cat
 {
     
+    /** @var bool $declawed */
+    private $declawed;
```
Comment (ybelenko) [inline samples/server/petstore/php-slim/lib/Model/Cat.php]: Slim generator has been also affected. Duplicated model vars.

File: samples/server/petstore/php-slim/lib/Model/Dog.php [hunk 4]
```diff
@@ -10,6 +10,9 @@
 class Dog
 {
     
+    /** @var string $breed */
+    private $breed;
```
Comment (ybelenko) [inline samples/server/petstore/php-slim/lib/Model/Dog.php]: And here, duplicate of variable `$breed`.

Comment (wing328) [reply samples/server/petstore/php-slim/lib/Model/Dog.php]: @wing328 yup, I have a fix to clean up the duplicates. Thanks for spotting the issue.

**Comment by YeTingGe** [comment]
Really glad to see we start working on this.....

**Comment by YeTingGe** [comment]
From the code and testing result, I suppose none of these (allOf, oneOf, anyOf) are supported now.

**Comment by wing328** [comment]
@YeTingGe we'll need to update the generators' templates to support the new mustache tags introduced by this PR, e.g. {{anyOf}}

Which generator(s) are you interested in? We can work with you to update those templates.

**Comment by JD-CSTx** [comment]
Add me to the list with YeTingGe. I too need a Java CodeGen implementation of anyOf, oneOf.

**Comment by YeTingGe** [comment]
> @YeTingGe we'll need to update the generators' templates to support the new mustache tags introduced by this PR, e.g. {{anyOf}}
> 
> Which generator(s) are you interested in? We can work with you to update those templates.

Could you please share your plan of adding java support for those fields? And also let me know how I help.

**Comment by hellboy81** [comment]
Waiting until ´anyOf´´ for Java/OpenFeign implemented

**Comment by jvans** [comment]
Add me to the list with YeTingGe. I also need a Java (Jax RS) CodeGen implementation of anyOf, oneOf.

**Comment by hellboy81** [comment]
> Waiting for anyOf, oneOf for Java as well.

@jvans Have you Gitter or Telegram?

**Comment by jonschoning** [comment]
regarding `haskell-http-client` generator

I've pushed an update to the branch which i think will fix it.

I was using `{{requiredVars}}` in my template.
It appears there were two behavior changes regarding  `{{requiredVars}}`

1. the `vendorExtensions` i had been previously using were not present
2. the `name` property of the `{{requiredVars}}` was different than the the `name` property of `{{#vars}}{{#required}}`

I have switched the code in this PR to avoid using `{{requiredVars}}` and switched to only use `{{#vars}}{{#required}}`

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java [hunk 5]
```diff
@@ -1329,16 +1330,53 @@ public String toDefaultValueWithParam(String name, Schema schema) {
...
+                            names.add(getAlias(schemaName));
+                        } else {
+                            LOGGER.warn("Error obtaining the datatype from ref:" + schema.get$ref() + ". Default to 'object'");
+                            return "object";
+                        }
+                    } else {
+                        // primitive type or model
+                        names.add(getAlias(getPrimitiveType(s)));
+                    }
+                    return "oneOf<" + String.join(",", names) + ">";
```
Comment (jonschoning) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java]: I'm guessing non-java langs will have to override this

**Comment by mm-matthias** [comment]
FWIW, I tested with some simple specs using allOf using the python generator. It works really well so far. I ran into the limitation of "allOf"/ComposedSchemas not begin supported, but this is clearly logged.
I'm also writing my own codegen against the oas3-spec-support1 branch and things have worked out very nice as well.
Thanks for adding this sorely needed feature!

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java (hunk 5 above)
Comment (wing328) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java]: I think one way is to override getSchemaType in the generator and then parse the string, e.g.
```
String result = super.getSchemaType(schema);
// parse the string here e.g. oneOf<Cat, Dog>
```

**Comment by wing328** [comment]
@mm-matthias thanks for reviewing the change. Can you ping me via https://gitter.im (ID: wing328) or Google hangout (email address in my Github profile page) for a quick chat?

**Comment by wing328** [comment]
Let's go with this PR to start with. If anyone encounters issues related to allOf/anyOf/oneOf, please open an issue with the details.

**Comment by hellboy81** [comment]
> Waiting for anyOf, oneOf for Java as well.

+1000

**Comment by tomghyselinck** [comment]
I tried the latest build, but it looks like `allOf` is not supported for Python either yet.
We definitely need this too.

**Comment by wing328** [comment]
@tomghyselinck do you mind opening an issue and share the spec so that we can reproduce the issue more easily?

**Comment by tomghyselinck** [comment]
@wing328 I stripped down my pretty huge `yaml` file and file BUG report #1657

**Comment by marcelstoer** [comment]
I would love to give this a try. Is there a binary somewhere that includes this fix? Specifically I want to find out if this addresses the v2 Java multi-level hierarchy bug (not sure it was even reported) where it would not generate the correct Jackson `@JsonSubTypes` for
```
C allOf B allOf A
```
on A. That's the only reason I'm currently still stuck on Swagger gen 2.4; would love to migrate.

**Comment by wing328** [comment]
@marcelstoer we only added the mustache tags for oneOf, anyOf and allOf but have not yet updated the templates for all the languages/generators to take advantage of these new features in OAS v3.

For Java client, please open a new issue with a spec and the current vs expected output for tracking.

**Comment by wing328** [comment]
@YeTingGe are you free for a quick chat via https://gitter.im (ID: wing328) about your use case?

**Comment by smasala** [comment]
@wing328 TypeScript generator has a weird side effect in which the interface property typing is prefixed with the word `AnyOf` or `OneOf`.

```
import {AnyOfMyType} from '../models/anyOfMyType.ts';   // doesn't exists. Only exists without the prefix
 
interface MyInt {
     myProp:  AnyOfMyType | null;
}
```

We need to define `anyOf | oneOf` so that we can use a `$ref` in conjunction with `nullable: true` https://github.com/OAI/OpenAPI-Specification/issues/1368

**Comment by strobox** [comment]
I found interesting part in ReDoc petstore.yaml OpenAPI 3.0 sample:
```yaml
  requestBodies:
    Pet:
      content:
        application/json:
          schema:
            allOf:
              - description: My Pet
                title: Pettie
              - $ref: '#/components/schemas/Pet'
```
I think it is some kind of inline overriding of default models using *allOf* notation

**Comment by spacether** [comment]
> Fix `{{{#vars}{{{/vars}}}` codegen array, please.
#vars does not include required variables on 2019-01-23

**Comment by jakzal** [comment]
Am I right to think this has only been implemented for ruby?

**Comment by spacether** [comment]
Composition in the root level or in deeper inline levels is working in [python-experimental](https://github.com/OpenAPITools/openapi-generator/blob/master/docs/generators/python-experimental.md#metadata) Here is a PR that shows inline composition working https://github.com/OpenAPITools/openapi-generator/pull/11420

**Pull request merged**